src/Diagnostics/CodeAnalysis/Core/MetaAnalyzers/RegisterActionAnalyzer.cs

```diff
@@ -245,10 +371,172 @@ protected override void AnalyzeNode(SymbolAnalysisContext symbolContext, TInvoca
+                var contextParameter = model.GetSymbolInfo(receiver, cancellationToken).Symbol as IParameterSymbol;
+                if (contextParameter == null)
+                {
+                    return;
+                }
+                
+                // Check if we are bailing out on this context parameter.
+                if (_startAnalysisContextParamsToSkip != null && _startAnalysisContextParamsToSkip.Contains(contextParameter))
+                {
+                    return;
+                }
+
+                _nestedActionsMap = _nestedActionsMap ?? new Dictionary<IParameterSymbol, List<NodeAndSymbol>>();
+                List<NodeAndSymbol> registerInvocations;
+                if (!_nestedActionsMap.TryGetValue(contextParameter, out registerInvocations))
+                {
+                    registerInvocations = new List<NodeAndSymbol>();
+                }
+
+                registerInvocations.Add(new NodeAndSymbol { Invocation = invocation, Method = method });
+                _nestedActionsMap[contextParameter] = registerInvocations;
+            }
+
+            private void CodeBlockEndAction(CodeBlockAnalysisContext codeBlockContext)
+            {
+                if (_declaredStartAnalysisContextParams == null)
+                {
+                    // No declared context parameters to analyze.
+                    return;
+                }
+
+                foreach (var contextParameter in _declaredStartAnalysisContextParams)
+                {
+                    // Check if we must bail out on this context parameter.
+                    if (_startAnalysisContextParamsToSkip != null && _startAnalysisContextParamsToSkip.Contains(contextParameter))
+                    {
+                        continue;
+                    }
+
+                    var hasEndAction = false;
+                    var hasNonEndAction = false;
+
+                    List<NodeAndSymbol> registeredActions = null;
+                    if (_nestedActionsMap != null && _nestedActionsMap.TryGetValue(contextParameter, out registeredActions))
+                    {
+                        foreach (var invocationInfo in registeredActions)
+                        {
+                            switch (invocationInfo.Method.Name)
+                            {
+                                case RegisterCompilationEndActionName:
+                                case RegisterCodeBlockEndActionName:
+                                    hasEndAction = true;
+                                    break;
+
+                                default:
+                                    hasNonEndAction = true;
+                                    break;
+                            }
+                        }
+                    }
+
+                    // Report a diagnostic if no non-end actions were registered on start analysis context parameter.
+                    if (!hasNonEndAction)
```

Comment:
Can this be fooled by cross-method registration? For example, if a code block start action calls a method that registers a syntax node action (rather than registering the syntax node action directly), will this analyzer complain? (This analyzer is a good thing to have even if the answer is "yes", but that would make it especially important that the diagnostics be gently suggestive rather than accusatory.)